### PR TITLE
Add support for Clipboard in VNC

### LINKED ISSUE #10306: Support NoVNC clipboard copy paste feature.
I use noVNC html version to connect to kubevirt VM.

And when I connect to Windows VM. Is there any way to support noVNC copy paste feature?

### PR DESCRIPTION
**Pull request by victortoso**
**What this PR does / why we need it**:

Clipboard copy-paste is a very useful feature when the user needs to interact between Remote VM OS and Client OS (local).

QEMU supports this feature since [v6.1.0](https://gitlab.com/qemu-project/qemu/-/commit/0bf41cab93e5c72dcda717abd625698b59d9ba3e) and libvirt since [v8.4.0](https://gitlab.com/libvirt/libvirt/-/commit/05b09f039ea6547efc39c2c483dbb68ad5265f13).

We just need to create a new virtio channel for guest-host communication.

The VNC client should support [Extended Clipboard Pseudo-Encoding](https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#extended-clipboard-pseudo-encoding). Note that gtk-vnc [does not support it](https://gitlab.gnome.org/GNOME/gtk-vnc/-/issues/22), so remote-viewer or gtk-vnc depended clients are not going to work. I tested with [TigerVNC](https://tigervnc.org/)

On the Guest side, we need spice-vdagent installed for the clipboard delivery. It is normally installed by default in most Linux distributions. On Windows, you can fetch the MSI installer at SPICE's [download](https://www.spice-space.org/download.html) page.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #10306

**Special notes for your reviewer**:

At the moment, Clipboard in QEMU does not support for migration.
See: https://gitlab.com/qemu-project/qemu/-/issues/1380

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
Enable VNC Clipboard
```

### CODE REVIEW

**Comment by victortoso** [comment]
> {"component":"virt-launcher","kind":"","level":"error","msg":"Live migration failed.","name":"testvmi-smk52","namespace":"kubevirt-test-default1","pos":"live-migration-source.go:969","reason":"error encountered during MigrateToURI3 libvirt api call: virError(Code=55, Domain=10, Message='Requested operation is not valid: cannot migrate domain: The vdagent chardev doesn't yet support migration')","timestamp":"2024-01-05T18:14:57.316247Z","uid":"750c25ae-9455-4cb9-9608-0480ad9c3e35"}

Ah, qemu vdagent does not support migration yet. Bummer.
https://gitlab.com/qemu-project/qemu/-/issues/1380

Considering that live migration is [not meant](https://github.com/kubevirt/kubevirt/pull/8047) to be behind a feature gate anymore, this feature might need another a feature gate (which inhibits migration) or should wait till qemu and libvirt implement migration to the feature.

**Review by vasiliy-ul — COMMENTED** [review]
From the code perspective, I do not see any issues, it looks good. However, I am somewhat uncertain about the proposed API `vmi.Spec.Domain.Desktop.Clipboard`. We already have `vmi.Spec.Domain.Devices.AutoattachGraphicsDevice` field to enable/disable VNC:

https://github.com/kubevirt/kubevirt/blob/afef58cb6e23b096d5e7a484e44f4689f715e832/staging/src/kubevirt.io/api/core/v1/schema.go#L458-L460

And in this PR the filed controlling the clipboard (which is tightly coupled to VNC) is put somewhere aside, in a new `VirtualDesktop` struct. Personally for me this does not look very intuitive from a user POV. Maybe we should think about grouping the APIs together somehow? Any thoughts on that?

**Comment by victortoso** [comment]
> And in this PR the filed controlling the clipboard (which is tightly coupled to VNC) is put somewhere aside, in a new VirtualDesktop struct. Personally for me this does not look very intuitive from a user POV. 

My thought process was to start separating features related to VDI now. To be honest, I wanted to have `vmi.Spec.Domain.Devices.ClientPassthrough` here as well as both features, clipboard and remote usb, can be group together as VDI related.

The tricky part is if we are going to consider VDI use cases that are attached to the QEMU, like VNC is now, SPICE was in the past (I mean in [KubeVirt](https://github.com/kubevirt/kubevirt/pull/618)) and perhaps RDP [will be in the future](https://wiki.qemu.org/Google_Summer_of_Code_2023#RDP_server), in oppose to solutions that run in the Guest (both VNC/RDP can, but also non open source ones like Citrix, NICE, etc)

My idea behind having `vmi.Spec.Domain.Desktop` is to provide a way to toggle those features on/off and allow for ClusterAdmin to actually be able to restrict it too later on, if we need it.

> Maybe we should think about grouping the APIs together somehow? Any thoughts on that?

Right. I made a clear distinction between Remote Protocols (VNC, RDP) and its features (Clipboard, remote USB).
You are considering having a more focused VNC api today, correct?

We might still use `vmi.Spec.Domain.Desktop` for that (unless we should consider Desktop a device too?)
We could have a enum for supported remote desktop protocols there that enables/disables VNC graphics for example, making `vmi.Spec.Domain.Devices.AutoattachGraphicsDevice` redundant.
 
@alicefr @andreabolognani, suggestions?
cc also @xpivarc and @fabiand

**Comment by vasiliy-ul** [comment]
> My thought process was to start separating features related to VDI now. 
> You are considering having a more focused VNC api today, correct?

Well, tbh I was looking at the PR without having VDI in mind. Current VNC api is not designed for VDI usecases afaik. Do you already have some thoughts on a broader VDI api? Maybe we should discuss it separately somehow?

**Comment by victortoso** [comment]
>  Current VNC api is not designed for VDI usecases afaik

Indeed.

> Do you already have some thoughts on a broader VDI api?

When I suggested the `vmi.Spec.Domain.Desktop` was with VDI in mind. I worked with it in the past so I think i could not help myself with it, but your early statement is correct. I don't think KubeVirt has VDI in mind at the moment, so I should take a step back about it. 

> Maybe we should discuss it separately somehow?

Sure, I've started a thread in the google groups.
https://groups.google.com/g/kubevirt-dev/c/fRvVWVXTtZY

**Comment by victortoso** [comment]
The [discussion](https://groups.google.com/g/kubevirt-dev/c/fRvVWVXTtZY) went forward with some ideas but one problem with clipboard over VNC is the dependency of SPICE guest agent.

I don't think QEMU guest agent would get this feature any time soon, so let's close this PR for now.
Note that, user's can still use Sidecars to enable this. A small guide was provided in [this comment.](https://github.com/kubevirt/kubevirt/issues/10306#issuecomment-1881549873)

**Comment by fabiand** [comment]
@victortoso how on the qemu side could help with the guest agent side?

**Comment by victortoso** [comment]
> @victortoso how on the qemu side could help with the guest agent side?

wrt clipboard, qemu uses spice's guest agent called spice-vdagent. QEMU guest-agent has no role with clipboard.
I believe this is on purpose, because implementing clipboard support depends on guest's platform and I don't think anyone in QEMU is planning to work and maintain that. spice-vdagent itself is becoming outdated as it does not have wayland support at the moment.

@elmarco, wdyt? I wonder if that could be somewhat different with RDP?

**Comment by fabiand** [comment]
Let me rephrase, I was wondering if the VNC protocol has support or could be extended to support clipboard as well, and then extend qemu and qemu ga to leverage this new protocl support.

It actually seems as if Gerd has looked into this, and as if the VNC protocol does support plaintext: https://lore.kernel.org/all/20210202111709.GB4168502@redhat.com/T/

**Comment by victortoso** [comment]
> Let me rephrase, I was wondering if the VNC protocol has support or could be extended to support clipboard as well, and then extend qemu and qemu ga to leverage this new protocl support.

Yes, VNC has clipboard in the protocol, it only depends if the client/server implement them.
I think you are suggesting that QEMU could implement some option 2 of Gerd's email. That can be done, but is very limited, host to guest only (not two ways) and, from implementation point of view, @berrange might think this could be very very slow.

If we have a path here, I could investigate. I closed thie issue/PR because I don't see one :(
Let's see if Daniel or Marc-Andre have any suggestions for us.

**Comment by berrange** [comment]
> I think you are suggesting that QEMU could implement some option 2 of Gerd's email. That can be done, but is very limited, host to guest only (not two ways) and, from implementation point of view, @berrange might think this could be very very slow.

Not only is it slow, but it is unreliable. The emulated keyboards have a limited input buffer which is quite small - there's a limit to how quickly humans can physically press keys.  If you tell QEMU to inject data via the emulated keyboard we need to pause between injecting each characters to avoid overflowing the keyboard buffer. If the pause is too short then data will be lost, or worse, corrupt; if it is too long then "pasting" data will be painfully slow and cannot be interrupted once started either.

Injecting text data via the emulated keyboard is not a credible solution IMHO.

**Comment by fabiand** [comment]
I agree ob the kbd method.
But I'd assume that guests - such as Linux (where I know it exists) and
Windows - gave protocols for pushing content to the clipboard. I'd further
assume that qemu-ga could leverage such a protocol.

On Tuesday, April 16, 2024, Daniel Berrangé ***@***.***>
wrote:
> I think you are suggesting that QEMU could implement some option 2 of
Gerd's email. That can be done, but is very limited, host to guest only
(not two ways) and, from implementation point of view, @berrange might
think this could be very very slow.
>
> Not only is it slow, but it is unreliable. The emulated keyboards have a
limited input buffer which is quite small - there's a limit to how quickly
humans can physically press keys. If you tell QEMU to inject data via the
emulated keyboard we need to pause between injecting each characters to
avoid overflowing the keyboard buffer. If the pause is too short then data
will be lost, or worse, corrupt; if it is too long then "pasting" data will
be painfully slow and cannot be interrupted once started either.
>
> Injecting text data via the emulated keyboard is not a credible solution
IMHO.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.<
https://ci3.googleusercontent.com/meips/ADKq_NYRU-37AEx2WLvnljXSEKS3MM0elR8sTy57_U42lvHooTZK6LcFOufLU1UxbUXyll5ioItL6Nn3gaGGPH8JrjzfyzkNFJBcqmm2MsaxWABlFg89FzKMQGTSmEYyRSUbwS2DKbwaz02vawjxduS5F66jQbHoIc61DauO6S9dci53KZnhTCa8csy-RQs1PligpP30ZITX8_BBl1u9Q6N3DOt0anB7UbU8ntJzwmTsSslPrk6QxZaDp7w=s0-d-e1-ft#https://github.com/notifications/beacon/AAAUMAAWZAY72Z5GYIORIA3Y5VB6PA5CNFSM6AAAAABBON2UVCWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTT2YARBI.gif>Message
ID: ***@***.***>

**Comment by berrange** [comment]
> But I'd assume that guests - such as Linux (where I know it exists) and Windows - gave protocols for pushing content to the clipboard. I'd further assume that qemu-ga could leverage such a protocol.

Linux in general has no concept of a clipboard. That is something provided by the desktop environment, with different impls for X11 and Wayland. There is a single qemu-ga running as a privileged user (root), while there can be multiple desktop environments running as different users. SPICE had a separate agent that ran as the user to handle clipboard integration, which then used a central agent as its transport. I don't know why they chose that multiplexing approach, perhaps because virtio-serial is single endpoint. If we ran over vsock, you could have a clipboard server backend in the host, connected to directly by multiple clipboard clients in the guest. IOW, I'm not convinced it makes conceptual sense to try to integrate this into `qemu-ga` at all.

**Comment by fabiand** [comment]
You are right. I was conflating too many things.

Wayland has a clipboard protocol, for sessions. qemu-ga is not a session related proc.
